rust/src/operations/transaction.rs

```diff
@@ -62,34 +62,41 @@ fn log_entry_from_actions(actions: &[Action]) -> Result<String, TransactionError
     Ok(jsons.join("\n"))
 }
 
-/// Low-level transaction API. Creates a temporary commit file. Once created,
-/// the transaction object could be dropped and the actual commit could be executed
-/// with `DeltaTable.try_commit_transaction`.
-async fn prepare_commit(
-    storage: &DeltaObjectStore,
-    operation: DeltaOperation,
-    mut actions: Vec<Action>,
+pub(crate) fn get_commit_bytes(
+    operation: &DeltaOperation,
+    actions: &mut Vec<Action>,
     app_metadata: Option<Map<String, Value>>,
-) -> Result<Path, TransactionError> {
+) -> Result<bytes::Bytes, TransactionError> {
     if !actions.iter().any(|a| matches!(a, Action::commitInfo(..))) {
-        let mut commit_info = Map::<String, Value>::new();
-        commit_info.insert(
-            "timestamp".to_string(),
-            Value::Number(serde_json::Number::from(Utc::now().timestamp_millis())),
-        );
-        commit_info.insert(
+        let mut extra_info = Map::<String, Value>::new();
+        let mut commit_info = operation.get_commit_info();
+        commit_info.timestamp = Some(Utc::now().timestamp_millis());
+        extra_info.insert(
             "clientVersion".to_string(),
             Value::String(format!("delta-rs.{}", crate_version())),
         );
-        commit_info.append(&mut operation.get_commit_info());
         if let Some(mut meta) = app_metadata {
-            commit_info.append(&mut meta)
+            extra_info.append(&mut meta)
         }
+        commit_info.info = extra_info;
         actions.push(Action::commitInfo(commit_info));
     }
 
     // Serialize all actions that are part of this log entry.
-    let log_entry = bytes::Bytes::from(log_entry_from_actions(&actions)?);
+    Ok(bytes::Bytes::from(log_entry_from_actions(actions)?))
+}
+
+/// Low-level transaction API. Creates a temporary commit file. Once created,
+/// the transaction object could be dropped and the actual commit could be executed
+/// with `DeltaTable.try_commit_transaction`.
+pub(crate) async fn prepare_commit(
+    storage: &dyn ObjectStore,
+    operation: &DeltaOperation,
+    actions: &mut Vec<Action>,
```

Comment:
Why is this changed to a `&mut` instead of just an owned value?

Reply:
I extracted this from the optimistic commit PR where we are using this in a loop and need access to the actions afterwards. That said, having this mutable in that scenario is likely not the best idea, and we need it only to add the commit info. Changed is so we just take a borrowed value now.